File: Leetcode/Perfect_Squares.cc
// Solution 1 : DP. 
// Time : O(N^2), Space : O(N)
// C++로는 80ms인데, 똑같은 솔루션으로 Python에서는 400ms가 넘어 Time Limit Exceeded 발생함.

class Solution {
public:
    int numSquares(int n) {
        static std::vector<int> min_nums = {0};
        while(min_nums.size() <= n) {
            int min_num = INT_MAX, i = min_nums.size();
            for(int j = 1; j*j <= i; j++)
                min_num = min(1 + min_nums[i-j*j], min_num);
            min_nums.push_back(min_num);
        }
        return min_nums[n];
    }
    
    int numSquares_1(int n) {
        std::vector<int> min_nums = {0};
        for(int i = 1; i < n+1; i++) {
            int min_num = INT_MAX;
            for(int j = 1; j*j <= i; j++)
                min_num = min(1 + min_nums[i-j*j], min_num);
            min_nums.push_back(min_num);
        }
        return min_nums[n];
    }
};

File: Leetcode/Perfect_Squares.py
# Solution 1 : DP. 
# Time Limit Exceeded. Time : O(N^2), Space : O(N)

# Solution 2 : BFS. 
# Time : O(N^2), Space : O(N)

class Solution:    
    def numSquares_2(self, n: int) -> int:
        q, depth = {n}, 0
        while q:
            depth += 1
            temp = set()
            for i in q:
                for j in range(1, int(n**0.5+1)):
                    if i == j*j:
                        return depth
                    temp.add(i-j*j)
                q = temp
        return depth
    
    
    def numSquares_1(self, n: int) -> int:
        min_nums = [0]
        for i in range(1, n+1):
            min_num = float('inf')
            for j in range(1, int(i**0.5+1)):
                min_num = min(1 + min_nums[i-j**2], min_num)
            min_nums.append(min_num)
        return min_nums[n]
